internal/cephfs/controllerserver.go

```diff
@@ -298,3 +409,250 @@ func (cs *ControllerServer) ControllerExpandVolume(ctx context.Context, req *csi
 		NodeExpansionRequired: false,
 	}, nil
 }
+
+// CreateSnapshot creates the snapshot in backend and stores metadata
+// in store
+// nolint: gocyclo
+func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error) {
+	if err := cs.validateSnapshotReq(ctx, req); err != nil {
+		return nil, err
+	}
+	requestName := req.GetName()
+	// Existence and conflict checks
+	if acquired := cs.SnapshotLocks.TryAcquire(requestName); !acquired {
+		klog.Errorf(util.Log(ctx, util.SnapshotOperationAlreadyExistsFmt), requestName)
+		return nil, status.Errorf(codes.Aborted, util.SnapshotOperationAlreadyExistsFmt, requestName)
+	}
+	defer cs.SnapshotLocks.Release(requestName)
+
+	// TODO take lock on parent subvolume
+
+	sourceVolID := req.GetSourceVolumeId()
+	// Find the volume using the provided VolumeID
+	volOptions, vid, err := newVolumeOptionsFromVolID(ctx, sourceVolID, nil, req.GetSecrets())
+	if err != nil {
+		var epnf util.ErrPoolNotFound
+		if errors.As(err, &epnf) {
+			klog.Warningf(util.Log(ctx, "failed to get backend volume for %s: %v"), sourceVolID, err)
+			return nil, status.Error(codes.NotFound, err.Error())
+		}
+		var evnf ErrVolumeNotFound
+		if errors.As(err, &evnf) {
+			return nil, status.Error(codes.NotFound, err.Error())
+		}
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	// TODO check snapshot exists
+
+	volOptions.RequestName = req.GetName()
```

Comment:
It is best not to reuse the same volOptions and just change the request name interchageably with the snapshot name or clone volume name. We should operate on separate volOptions for each entity keeping this cleaner.